application/controllers/OptinController.php

```diff
@@ -42,7 +42,7 @@ function actiontokens($surveyid, $token, $langcode = '')
         if (!$iSurveyID) {
             $this->redirect(array('/'));
         }
-        $iSurveyID = (int) $iSurveyID;
+        $iSurveyID = $oSurvey->primaryKey;
```

Comment:
Maybe can remove this line : https://github.com/TonisOrmisson/LimeSurvey/blob/b8af9b7256deb22e8ed8d8b3b7f67b86652182cc/application/controllers/OptinController.php#L37 (directly use  `$oSurvey = Survey::model()->findByPk($surveyid);` )

And send a 404 error if $oSurvey is null